src/runtime/qurt_hvx.cpp

```diff
@@ -52,4 +53,214 @@ WEAK void halide_qurt_hvx_unlock_as_destructor(void *user_context, void * /*obj*
+
+#ifdef DEBUG_PREFETCH // Prefetch debug enabled
+
+// Only produce debug info in this range
+#define DBG_P_START          1  // First debug instance to display
+#define DBG_P_STOP           9  // Last debug instance to display
+//#define DBG_P_STOP           0xFFFFFFFF  // Last debug instance to display
+
+#define DBG_PREFETCH_V(...)  log_printf(__VA_ARGS__) // Verbose debug enabled
+//#define DBG_PREFETCH_V(...)  {} // Verbose debug disabled
+#define DBG_PREFETCH(...)    if ((dbg_cnt >= DBG_P_START) && \
+                                 (dbg_cnt <= DBG_P_STOP)) { \
+                                 log_printf(__VA_ARGS__); \
+                             }  // Normal range-based debug enabled
+
+#else // Prefetch debug disabled
+
+#define DBG_PREFETCH_V(...)  {} // Verbose debug disabled
+#define DBG_PREFETCH(...)    {} // Normal debug disabled
+
+#endif
+//
+WEAK int halide_hexagon_prefetch_buffer_t(const uint32_t dim, const int32_t elem_size,
+                const uint32_t num_elem, const buffer_t *buf)
+{
+    // Extract needed fields from buffer_t
+    unsigned char *addr   = buf->host;
+    const int32_t *extent = buf->extent;
+    const int32_t *stride = buf->stride;
+    const int32_t *min    = buf->min;
+    const uint32_t boxdim = dim;  // dimensions of entire box to prefetch
+    const unsigned char *addr_beg = addr;
+    const unsigned char *addr_end = addr + (num_elem * elem_size);
+
+#ifdef DEBUG_PREFETCH
+    static uint32_t dbg_cnt = 0;  // limit how many calls debug is generated
+    dbg_cnt++;
+    DBG_PREFETCH("halide_hexagon_prefetch_buffer_t(%u, %d)\n", boxdim, elem_size, num_elem);
+
+    DBG_PREFETCH("  addr range 0x%p => 0x%p\n", addr_beg, addr_end);
+
+    DBG_PREFETCH("  buf host=0x%p elem_size=%d\n", addr, buf->elem_size);
+    for (uint32_t i = 0; i < boxdim; i++) {
+        DBG_PREFETCH("  buf stride[%d]=0x%-8x min[%d]=%-6d ext[%d]=%d\n",
+                          i, stride[i], i, min[i], i, extent[i]);
+    }
+#endif
+
+    // Compute starting position of box
+    int32_t startpos = 0;
+    for (uint32_t i = 0; i < boxdim; i++) {
+        startpos += min[i] * stride[i];
+    }
+    addr += startpos * elem_size;
+    DBG_PREFETCH("  +startpos=0x%x*%d => addr=0x%p\n", startpos, elem_size, addr);
+
+    // Range check starting address
+    if ((addr < addr_beg) || (addr >= addr_end)) {
+        DBG_PREFETCH_V("  l2fetch: 0x%p out of range [0x%p, 0x%p]\n",
+                        addr, addr_beg, addr_end);
+        return 0;
+    }
+
+    // Compute 2-D prefetch descriptor
+    // l2fetch(Rs,Rtt): 48 bit descriptor
+    uint32_t pdir    = 1;       // Prefetch direction: 0=rows, 1=columns
+    uint32_t pstride = stride[0] * elem_size;
+    uint32_t pwidth  = extent[0] * elem_size;
+    uint32_t pheight = 1;       // Initially 1-D descriptor
+
+    uint32_t iterdim = 1;       // Dimension to iterate over
+    if (boxdim > 1) {           // Potential for 2-D descriptor
```

Comment:
The prefetch descriptor itself is only the inner 2 dimensions.  Higher dimensions are supported by iterating over a series of 2-D prefetches.